File: 01_two_pointers/04_trapping_rain_water/trapping_rain_water.py
from typing import List

# Reference: https://www.youtube.com/watch?v=ZI2z5pq0TqA

class Solution:
    def trap(self, height: List[int]) -> int:
        l = 0
        r = len(height) - 1
        l_max = 0
        r_max = 0
        volume = 0
        while l <= r:
            if l_max <= r_max:
                water_in_column = min(l_max, r_max) - height[l]
                if water_in_column > 0:
                    volume += water_in_column
                if height[l] > l_max:
                    l_max = height[l] 
                l += 1
            else:
                water_in_column = min(l_max, r_max) - height[r]
                if water_in_column > 0:
                    volume += water_in_column
                if height[r] > r_max:
                    r_max = height[r] 
                r -= 1
                
        return volume

if __name__ == "__main__":
    solution = Solution()
    print(solution.trap([0, 1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1]), ": should be 6")
    print(solution.trap([4,2,0,3,2,5]), ": should be 9")
    print(solution.trap([4,2,3]))
    print(solution.trap([5,4,1,2]))
    print(solution.trap([4,9,4,5,3,2]))
    

File: NeetCode/01_two_pointers/04_trapping_rain_water/trapping_rain_water.cpp
#include <vector>
#include <cstddef>

using std::vector;

class Solution
{
  public:
    int trap(vector<int>& height)
    {
        size_t l = 0;
        size_t r = height.size() - 1;

        int l_max = 0;
        int r_max = 0;

        int volumn = 0;
        while (l <= r)
        {
            if (l_max <= r_max)
            {
                int this_column = std::min(l_max, r_max) - height[l];
                if (this_column > 0)
                    volumn += this_column;
                if (height[l] > l_max)
                    l_max = height[l];
                ++l;
            }
            else
            {
                int this_column = std::min(l_max, r_max) - height[r];
                if (this_column > 0)
                    volumn += this_column;
                if (height[r] > r_max)
                    r_max = height[r];
                --r;
            }
        }
        return volumn;
    }
};

int main()
{
    // int ary[] = {0, 1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1};
    int ary[] = {4, 2, 3};

    std::vector<int> vec;
    for (size_t i = 0; i < sizeof(ary) / sizeof(ary[0]); ++i)
        vec.push_back(ary[i]);

    Solution solution;
    solution.trap(vec);
}